main/src/library/StringBuilder.flix

```diff
@@ -162,17 +162,17 @@ namespace StringBuilder {
     ///
     /// Return the length of the StringBuilder `sb`.
     ///
-    pub def length(sb: StringBuilder): Int32 =
+    pub def length(sb: StringBuilder): Int32 & Impure =
         import java.lang.StringBuilder.length();
         let StringBuilder(msb) = sb;
-        msb.length() as & Pure
+        msb.length()
 
     ///
     /// Convert the StringBuilder `sb` to a string.
     ///
-    pub def toString(sb: StringBuilder): String =
+    pub def toString(sb: StringBuilder): String & Impure =
         import java.lang.StringBuilder.toString();
         let StringBuilder(msb) = sb;
-        msb.toString() as & Pure
+        msb.toString()
 
```

Comment:
OK; I see now that I have made a mistake. What I said earlier was simply wrong:

We must leave `length` and `toString` of StringBuilder as impure. The problem is the same as always. This:

```
let sb = StringBuilder.new();
let l = StringBuilder.length(sb);
sb.append(...);
l
```

is *not* equivalent to:

```
let sb = StringBuilder.new();
sb.append(...);
StringBuilder.length(sb)
```

This is the smoke test: Purity must permit equational reasoning.

Now that said -- and perhaps the reason I was confused -- is that it is entirely possible for a function that uses a string builder to be pure (even if it is impure on the inside). I think you already got this because I see that you use this pattern correctly in e.g. `unfoldWithIter` (and many others).

Reply:
Yes that pattern is very much like `runST` in Haskell which has been very helpful to think about when considering effects.